server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java

```diff
@@ -229,39 +230,48 @@ private FieldCapabilitiesResponse merge(
         FieldCapabilitiesRequest request,
         List<FieldCapabilitiesFailure> failures
     ) {
-        final List<FieldCapabilitiesIndexResponse> indexResponses = indexResponsesMap.values()
-            .stream()
-            .sorted(Comparator.comparing(FieldCapabilitiesIndexResponse::getIndexName))
-            .toList();
-        final String[] indices = indexResponses.stream().map(FieldCapabilitiesIndexResponse::getIndexName).toArray(String[]::new);
-        final Map<String, Map<String, FieldCapabilities.Builder>> responseMapBuilder = new HashMap<>();
-        for (FieldCapabilitiesIndexResponse response : indexResponses) {
+        Map<String, FieldCapabilitiesIndexResponse> responses = new TreeMap<>(indexResponsesMap);
+        Map<String, Map<String, FieldCapabilities.Builder>> responseMapBuilder = new HashMap<>();
+        for (FieldCapabilitiesIndexResponse response : responses.values()) {
             innerMerge(responseMapBuilder, request, response);
         }
-        final Map<String, Map<String, FieldCapabilities>> responseMap = new HashMap<>();
+        Map<String, Map<String, FieldCapabilities>> responseMap = new HashMap<>();
+        // these may be referenced a lot - keep a canonical object around
+        Map<String, FieldCapabilities.IndexCaps> unmappedIndexCaps = new HashMap<>();
+
         for (Map.Entry<String, Map<String, FieldCapabilities.Builder>> entry : responseMapBuilder.entrySet()) {
             final Map<String, FieldCapabilities.Builder> typeMapBuilder = entry.getValue();
             if (request.includeUnmapped()) {
-                addUnmappedFields(indices, entry.getKey(), typeMapBuilder);
+                addUnmappedFields(responses.keySet(), entry.getKey(), typeMapBuilder, unmappedIndexCaps);
             }
             boolean multiTypes = typeMapBuilder.size() > 1;
-            final Map<String, FieldCapabilities> typeMap = new HashMap<>();
-            for (Map.Entry<String, FieldCapabilities.Builder> fieldEntry : typeMapBuilder.entrySet()) {
-                typeMap.put(fieldEntry.getKey(), fieldEntry.getValue().build(multiTypes));
-            }
-            responseMap.put(entry.getKey(), Collections.unmodifiableMap(typeMap));
+            responseMap.put(
+                entry.getKey(),
+                Collections.unmodifiableMap(
```

Comment:
Done it this way rather than using `Collectors.toUnmodifiableMap`, as the collector copies the data into an unmodifiable collection, rather than just wrapping it

Reply:
I wonder if we can avoid some of this by replacing the `FieldCapabilities.Builder` here with a `Function<Boolean, FieldCapabilities>`, or perhaps making `Builder` an interface and having mapped and unmapped implementations.  The unmapped indices collection is basically a set of strings so that's all we need to record and we can then produce a `FieldCapabilities` object of type `unmapped` which lists all those indices as unsearchable/unaggregatable.  That way we can avoid leaking the `IndexCaps` record any further.

Reply:
I agree it's not great exposing `IndexCaps` outside the class, but it's tricky to manage the lifecycle of the `IndexCaps` objects (which have to be created at some point) just within the `FieldCapabilities` class, as there's lots of those managed independently by the `TransportFieldCapabilitiesAction` class.

I've done the simple thing of creating a class in `FieldCapabilities` that does this wrapping instead, how does that look?

Reply:
> IndexCaps objects (which have to be created at some point)

I don't think they do though? They're an implementation detail of `FieldCapabilities.Builder`.  And if we make `typeMapBuilder` a `Map<String, Function<Boolean, FieldCapabilities>>` then we can have an 'unmapped' implementation that takes a list of indices and returns a `FieldCapabilities` object called `unmapped` that lists them the indices in its `nonSearchable` and `nonAggregatable` fields.  No need to create `IndexCaps` objects, no need to share state.

Reply:
Aaah, I get what you mean now. I've updated it to build a `FieldCapabilities` directly